engine-precompiles/src/alt_bn256.rs

```diff
@@ -75,6 +75,7 @@ trait HostFnEncode {
 
 #[cfg(feature = "contract")]
 fn concat_low_high<const P: usize, const S: usize>(low: [u8; P], high: [u8; P]) -> [u8; S] {
+    assert!(S >= (P * 2), "INSUFFICIENT_OUTPUT_ARRAY_SIZE");
```

Comment:
For what it's worth, this could be done as a compile-time check instead since `P` and `S` are const generics:

```rust
#[cfg(feature = "contract")]
mod type_arith {
    pub struct Double<const P: usize>;
    pub trait Is<const S: usize> {}
    impl Is<128> for Double<64> {}
    impl Is<64> for Double<32> {}
    impl Is<32> for Double<16> {}
}

#[cfg(feature = "contract")]
fn concat_low_high<const P: usize, const S: usize>(low: [u8; P], high: [u8; P]) -> [u8; S]
where
    type_arith::Double<P>: type_arith::Is<S>,
{
    let mut bytes = [0u8; S];
    bytes[0..P].copy_from_slice(&low);
    bytes[P..P * 2].copy_from_slice(&high);
    bytes
}
```

I think the compile-time approach has two advantages over the runtime one: (a) no panics, (b) no performance impact. The disadvantage is that the code is a little harder to read. Personally I think it is worth the slightly arcane looking code in case.

Reply:
There is [a more elegant way](https://crates.io/crates/const_guards), but it requires an unstable language feature (and I know we are trying to get off of nightly).

Reply:
Yes. The coming soon `1.68.0` stabilises `alloc_error_handler`, but `PanicInfo::message(&self)`  is still under the `panic_info_message` feature, unfortunately.